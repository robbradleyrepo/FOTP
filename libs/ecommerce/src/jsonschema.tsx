import React, { FC } from "react";
import { JsonLd } from "react-schemaorg";
import { Product as ProductDts } from "schema-dts";

import { getLegacyId } from "./common";
import { getProductUrl, getProductVariants, Product } from "./product";

interface ProductSchemaProps {
  product: Product;
}

type GtinDts = Pick<ProductDts, "gtin8" | "gtin12" | "gtin13" | "gtin14">;

const getGtinData = (value: string | null): GtinDts | undefined => {
  const GTIN_LENS = [8, 12, 13, 14];

  if (!value || !GTIN_LENS.includes(value.length)) {
    // Invalid value
    return;
  }

  // Seems strange that you need to be explicit with the gtin length value
  // rather than use the generic `gtin` value.

  const key = `gtin${value.length}` as "gtin8" | "gtin12" | "gtin13" | "gtin14";
  return { [key]: value };
};

export const useProductSchemaJson = (product: Product) => {
  const variants = getProductVariants(product);
  const defaultVariant = variants[0];
  const name = `${product.title}${
    product.subtitle?.value ? " â€“ " + product.subtitle.value : ""
  }`.trim();
  // You can just set the expiry date to next year and have it
  // dynamically updated every day.
  const priceValidUntil = new Date(
    new Date().setFullYear(new Date().getFullYear() + 1)
  )
    .toISOString()
    .substring(0, 10);
  const url = getProductUrl(product);

  return {
    "@type": "Product",
    aggregateRating:
      product.bottomline && product.bottomline.totalReviews > 0
        ? {
            "@type": "AggregateRating",
            ratingCount: product.bottomline.totalReviews,
            ratingValue: product.bottomline.averageScore,
          }
        : undefined,
    brand: product.vendor,
    description: product.description,
    ...getGtinData(defaultVariant.barcode),
    image:
      product.images.edges.length > 0
        ? [product.images.edges[0].node.url]
        : undefined,
    name,
    offers: variants.map((variant) => {
      return {
        "@type": "Offer",
        availability: variant.availableForSale
          ? "https://schema.org/InStock"
          : "https://schema.org/OutOfStock",
        description: product.description,
        ...getGtinData(variant.barcode),
        image: variant.image.url,
        itemCondition: "https://schema.org/NewCondition",
        name:
          variant.title !== "Default Title"
            ? `${name} - ${variant.title}`
            : name,
        price: variant.priceV2.amount,
        priceCurrency: variant.priceV2.currencyCode,
        priceValidUntil,
        // As a DTC, vendor and seller is always the same.
        seller: product.vendor,
        sku: variant.sku,
        url: `${url}?variant=${getLegacyId(variant.id)}`,
      };
    }),
    productID: getLegacyId(product.id),
    sku: defaultVariant.sku,
    url,
  };
};

export const ProductSchema: FC<ProductSchemaProps> = ({ product }) => {
  const productSchemaJson = useProductSchemaJson(product);

  return (
    <JsonLd<ProductDts>
      item={{
        "@context": "https://schema.org",
        // Cast to `ProductDts`, as it's better to have the inferred types on
        // `useProductSchemaJson` and it's difficult to narrow the types from
        // `schema-dts`
        ...(productSchemaJson as ProductDts),
      }}
    />
  );
};
