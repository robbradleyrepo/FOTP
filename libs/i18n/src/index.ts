export { I18nHead, I18nSwitch, I18nSwitcher } from "./components";
export { configureI18n, getI18nInstance } from "./config";
export { LocaleProvider, useLocale } from "./context";
export { useCurrencyFormatter, useDateTimeFormatter } from "./formatters";
export {
  findLocaleByLangtag,
  findLocaleByNextLocale,
  findLocaleByQuery,
} from "./helpers";
export type { I18nConfig, Locale } from "./types";
