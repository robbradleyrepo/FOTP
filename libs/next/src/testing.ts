import { Router } from "next/router";
import { ParsedUrlQuery } from "querystring";
import { mocked } from "ts-jest/utils";

import { getAsPath } from "./helpers";

export const getMockedRouter = (pathname: string, params?: ParsedUrlQuery) => {
  const asPath = getAsPath(pathname, params);

  return mocked(({
    asPath,
    events: {
      off: jest.fn(),
      on: jest.fn(),
    },
    isReady: true,
    pathname,
    prefetch: jest.fn(async () => undefined),
    push: jest.fn(),
    query: params,
    replace: jest.fn(),
    route: pathname,
  } as unknown) as Router);
};
