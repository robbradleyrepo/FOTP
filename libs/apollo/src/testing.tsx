import { MockedResponse, MockLink } from "@apollo/react-testing";
import { ApolloClientType, NextPageWithApollo } from "@sss/apollo";
import { Locale } from "@sss/i18n";
import { getMockedRouter } from "@sss/next/testing";
import {
  act,
  render,
  RenderOptions,
  RenderResult,
} from "@testing-library/react";
import {
  InMemoryCache,
  InMemoryCacheConfig,
  IntrospectionFragmentMatcher,
  IntrospectionResultData,
} from "apollo-cache-inmemory";
import { ApolloClient } from "apollo-client";
import {
  GetServerSideProps,
  GetServerSidePropsContext,
  GetStaticProps,
} from "next";
import { RouterContext } from "next/dist/shared/lib/router-context";
import { Router } from "next/router";
import { ParsedUrlQuery, stringify } from "querystring";
import React, { ComponentType } from "react";

import { initializeApollo } from "./client";
jest.mock("./client");

interface BaseSandbox {
  // The global mocks to apply
  globalMocks?: MockedResponse[];
  // Additional mocks under test
  mocks?: MockedResponse[];
  renderOptions?: Omit<RenderOptions, "queries">;
  pathname?: string;
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  params?: { [key: string]: any };
}

interface PageSandbox extends BaseSandbox {
  NextPage: {
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    default: React.ComponentType<any>;
    getStaticProps?: GetStaticProps;
    getServerSideProps?: GetServerSideProps;
  };
}

interface PageResult {
  mockLink: MockLink;
  redirect?: { permanent: boolean; destination: string };
  notFound?: boolean;
  page: RenderResult;
  router: Router;
}

interface ComponentSandbox extends BaseSandbox {
  Component: React.ReactNode;
}

interface ComponentResult {
  mockLink: MockLink;
  component: RenderResult;
  router: Router;
}

interface SandboxInit {
  globalMocks?: MockedResponse[];
  introspectionResultData?: IntrospectionResultData;
  locale: Locale;
  Wrapper: ComponentType<{
    apolloClient: ApolloClientType;
    locale: Locale;
  }>;
}

export const createApolloClient = (
  mocks: MockedResponse[],
  introspectionQueryResultData?: IntrospectionResultData
) => {
  const cacheConfig: InMemoryCacheConfig = { addTypename: true };

  // `InMemoryCache` doesn't like `fragmentMatcher: undefined`. We must omit it.
  if (introspectionQueryResultData) {
    cacheConfig.fragmentMatcher = new IntrospectionFragmentMatcher({
      introspectionQueryResultData,
    });
  }

  return new ApolloClient({
    cache: new InMemoryCache(cacheConfig),
    link: new MockLink(mocks, true),
  });
};

export default class Sandbox {
  public constructor(private opts: SandboxInit) {}

  protected createApolloClient(
    mocks: MockedResponse[] = [],
    globalMocks: MockedResponse[] = this.opts.globalMocks ?? []
  ) {
    return createApolloClient(
      [...globalMocks, ...mocks],
      this.opts.introspectionResultData
    );
  }

  public async page({
    NextPage: NextPageBundle,
    globalMocks,
    mocks,
    renderOptions,
    pathname = "/",
    params = {},
  }: PageSandbox): Promise<PageResult> {
    const router = getMockedRouter(pathname, params);

    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    (global as any).window = Object.create(window);
    Object.defineProperty(window, "location", {
      value: {
        href: `https://localhost${router.asPath}`,
        pathname: router.asPath,
        search: params ? stringify(params) : "",
      },
      writable: true,
    });

    const apolloClient = this.createApolloClient(mocks, globalMocks);

    (initializeApollo as jest.Mock).mockReturnValueOnce(apolloClient);

    const { default: PageComponent } = NextPageBundle;

    const pageData = NextPageBundle.getStaticProps
      ? await NextPageBundle.getStaticProps({ locale: "en-US", params })
      : NextPageBundle.getServerSideProps
      ? await NextPageBundle.getServerSideProps(
          // `getServerSideProps` context argument also requires
          // `req: IncomingMessage` and `res: ServerResponse`. Cast
          // this for testing purposes.
          ({
            locale: "en-US",
            params,
            req: {},
            res: {
              setHeader: jest.fn(),
              statusCode: null,
            },
          } as unknown) as GetServerSidePropsContext<ParsedUrlQuery>
        )
      : { props: {} };

    let page;

    act(() => {
      const Wrapper = this.opts.Wrapper;
      page = render(
        "props" in pageData &&
          !("notFound" in pageData) &&
          !("redirect" in pageData) ? (
          <RouterContext.Provider value={router}>
            <Wrapper apolloClient={apolloClient} locale={this.opts.locale}>
              <PageComponent {...pageData.props} />
            </Wrapper>
          </RouterContext.Provider>
        ) : (
          <div />
        ),
        {
          ...renderOptions,
        }
      );
    });

    return {
      link: apolloClient.link,
      notFound: "notFound" in pageData && pageData.notFound,
      page,
      redirect: "redirect" in pageData && pageData.redirect,
      router,
    } as any; // eslint-disable-line @typescript-eslint/no-explicit-any
  }

  public async component({
    Component,
    ...options
  }: ComponentSandbox): Promise<ComponentResult> {
    const NextPageComponent: NextPageWithApollo = () => {
      return <>{Component}</>;
    };

    const { page: component, ...rest } = await this.page({
      NextPage: { default: NextPageComponent },
      ...options,
    });

    return {
      component,
      ...rest,
    };
  }
}
