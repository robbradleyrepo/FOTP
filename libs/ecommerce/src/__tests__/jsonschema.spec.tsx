import { configureI18n, LocaleProvider } from "@sss/i18n";
import { render } from "@testing-library/react";
import { advanceTo } from "jest-date-mock";
import React from "react";

import { ProductSchema } from "../jsonschema";
import { mockedResponses } from "../testing";

jest.mock("../config", () => ({
  __esModule: true,
  default: {
    customCheckout: false,
    origin: "https://example.com",
    subscriptions: false,
  },
}));

const locale = {
  countryCode: "US",
  label: "Looking for our US store?",
  langtag: "en-US",
  language: "en",
  region: "us",
  resource: {},
  timeZone: "America/Los_Angeles",
};

configureI18n({ fallbackLocale: locale, supportedLocales: [locale] });

describe("ProductSchema", () => {
  beforeEach(() => {
    advanceTo("2021-01-02");
  });

  it("should render a valid JSON", () => {
    const product = mockedResponses.product.productFragment;

    const root = render(
      <LocaleProvider locale={locale}>
        <ProductSchema product={product} />
      </LocaleProvider>
    );

    expect(root.asFragment()).toMatchSnapshot();
  });
});
