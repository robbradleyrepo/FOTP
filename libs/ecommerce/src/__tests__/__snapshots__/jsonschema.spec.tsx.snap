// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductSchema should render a valid JSON 1`] = `
<DocumentFragment>
  <script
    type="application/ld+json"
  >
    {"@context":"https://schema.org","@type":"Product","aggregateRating":{"@type":"AggregateRating","ratingCount":2,"ratingValue":4.5},"brand":"Front Of The Pack","description":"Targeted joint and mobility support","image":["https://cdn.shopify.com/mock-image/product-1.jpg"],"name":"Move – Targeted joint and mobility support","offers":[{"@type":"Offer","availability":"https://schema.org/InStock","description":"Targeted joint and mobility support","image":"https://cdn.shopify.com/mock-image/variant-4.jpg","itemCondition":"https://schema.org/NewCondition","name":"Move – Targeted joint and mobility support","price":"34.99","priceCurrency":"USD","priceValidUntil":"2022-01-02","seller":"Front Of The Pack","sku":"move","url":"https://example.com/products/move?variant=31746650406990"}],"productID":"4535537533006","sku":"move","url":"https://example.com/products/move"}
  </script>
</DocumentFragment>
`;
